### PR TITLE
Add RLHF guide and dummy demo with Keras/JAX

### PR DESCRIPTION
**Pull request by TrailChai**
This commit introduces a new example for Reinforcement Learning from Human Feedback (RLHF).

It includes:
- \`examples/rl/rlhf_dummy_demo.py\`: A Python script demonstrating a simple RLHF loop with a dummy environment, a policy model, and a reward model, using Keras with the JAX backend.
- \`examples/rl/md/rlhf_dummy_demo.md\`: A Markdown guide explaining the RLHF concept and the implementation details of the demo script.
- \`examples/rl/README.md\`: A new README for the RL examples section, now including the RLHF demo.

Note: The Python demo script (\`rlhf_dummy_demo.py\`) currently experiences timeout issues during the training loop in the development environment, even with significantly reduced computational load. This is documented in the guide and README. The code serves as a structural example of implementing the RLHF components.

### CODE REVIEW

File: examples/rl/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Keras Reinforcement Learning Examples
```
Comment (divyashreepathihalli) [inline examples/rl/README.md]: remove readme

File: examples/rl/md/rlhf_dummy_demo.md [hunk 2]
```diff
@@ -0,0 +1,234 @@
+# Reinforcement Learning from Human Feedback (RLHF) - Dummy Demo Guide
+
```
Comment (divyashreepathihalli) [inline examples/rl/md/rlhf_dummy_demo.md]: remove dummy from file name and title

File: examples/rl/md/rlhf_dummy_demo.md [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+This guide explains the concept of Reinforcement Learning from Human Feedback (RLHF) and walks through the components of the accompanying dummy demo script `rlhf_dummy_demo.py`.
+
+## 1. What is Reinforcement Learning from Human Feedback (RLHF)?
```
Comment (divyashreepathihalli) [inline examples/rl/md/rlhf_dummy_demo.md]: Since it is not really human feedback, but another model - should we call thsi RLAIF?

File: examples/rl/rlhf_dummy_demo.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    @jax.jit
+    def policy_loss_fn(policy_model_params, state_input, action, predicted_reward_value_stopped):
+        # stateless_call might return a tuple (e.g., (outputs, other_states) or just (outputs,))
+        # We are interested in the first element, which should be the main output tensor.
+        predictions_tuple = policy_model.stateless_call(
+            policy_model_params["trainable"], 
+            policy_model_params["non_trainable"], 
+            state_input
+        )
+        actual_predictions_tensor = predictions_tuple[0]
```
Comment (divyashreepathihalli) [inline examples/rl/rlhf_dummy_demo.py]: are we assuming batch size is 1?

File: examples/rl/rlhf_dummy_demo.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+            current_policy_loss, policy_grads_dict_step = policy_value_and_grad_fn(
+                policy_params_dict,
+                state_input_jax,
+                action_jax, # Use JAX array action
+                stopped_predicted_reward
+            )
+            episode_policy_losses.append(current_policy_loss)
+            policy_grads_step = policy_grads_dict_step["trainable"]
+            # Accumulate policy gradients
+            for i, grad in enumerate(policy_grads_step):
```
Comment (divyashreepathihalli) [inline examples/rl/rlhf_dummy_demo.py]: For potentially improving performance
`policy_grads_accum = jax.tree_map(lambda acc, new: acc + new if new is not None else acc, policy_grads_accum, policy_grads_step)`

File: examples/rl/md/rlhf_dummy_demo.md [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+    # JAX gradient functions using model.stateless_call
+    @jax.jit
+    def policy_loss_fn(policy_model_params, state_input, action, predicted_reward_value_stopped):
+        # ... (calculates policy loss based on predicted_reward_value_stopped)
+        predictions_tuple = policy_model.stateless_call(...)
+        actual_predictions_tensor = predictions_tuple[0]
+        action_probs = actual_predictions_tensor[0]
+        log_prob = jnp.log(action_probs[action] + 1e-7)
+        return -log_prob * predicted_reward_value_stopped
```
Comment (divyashreepathihalli) [inline examples/rl/md/rlhf_dummy_demo.md]: If this predicted_reward_value is just R(s,a), then it's using the immediate predicted reward.
it's a very naive and generally ineffective way to train a policy. Thee log_prob should be multiplied by the cumulative discounted future reward (Return, G_t).

**Review by divyashreepathihalli — COMMENTED** [review]
Thanks Yasir! just keep the .py file for now. once it is approved we can generate the .ipynb files

File: examples/rl/README.md (hunk 1 above)
Comment (TrailChai) [reply examples/rl/README.md]: done

File: examples/rl/md/rlhf_dummy_demo.md (hunk 2 above)
Comment (TrailChai) [reply examples/rl/md/rlhf_dummy_demo.md]: done

File: examples/rl/md/rlhf_dummy_demo.md (hunk 3 above)
Comment (TrailChai) [reply examples/rl/md/rlhf_dummy_demo.md]: Yes that makes sense. Thanks

File: examples/rl/rlhf_dummy_demo.py (hunk 4 above)
Comment (TrailChai) [reply examples/rl/rlhf_dummy_demo.py]: Yes. The code as written assumes a batch size of 1 for all model inputs and gradient calculations

File: examples/rl/rlhf_dummy_demo.py (hunk 5 above)
Comment (TrailChai) [reply examples/rl/rlhf_dummy_demo.py]: Thanks, I have refactored policy gradients and reward gradients accumulations using `jax.tree_map`.

File: examples/rl/rlhf_dummy_demo.py (hunk 4 above)
Comment (divyashreepathihalli) [reply examples/rl/rlhf_dummy_demo.py]: why is that?

**Comment by divyashreepathihalli** [comment]
the .md files are automatically generated. SO you might want to move the explanation content part to .py

File: examples/rl/md/rlhf_dummy_demo.md (hunk 6 above)
Comment (TrailChai) [reply examples/rl/md/rlhf_dummy_demo.md]: Commit https://github.com/keras-team/keras-io/pull/2117/commits/c212593788db6fa1265b162e0dc1c59641e61455 addresses this.

File: examples/rl/rlhf_dummy_demo.py (hunk 4 above)
Comment (TrailChai) [reply examples/rl/rlhf_dummy_demo.py]: This was a simplification for the demo's clarity and to manage complexity, especially since REINFORCE-style updates can be done on single trajectories. A more advanced setup would definitely use batching across multiple episodes or from a replay buffer for stability and efficiency.

Does that make sense in the context of this simplified demo?

**Comment by TrailChai** [comment]
I have deleted the .md files and added relevant documentation pieces to .py file
